model/Search/OpenSearchIndexer.php

```diff
@@ -0,0 +1,200 @@
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoAdvancedSearch\model\Search;
+
+use OpenSearch\Client;
+use Psr\Log\LoggerInterface;
+use RuntimeException;
+use Iterator;
+use oat\tao\model\search\index\IndexDocument;
+
+class OpenSearchIndexer implements IndexerInterface
+{
+    private const INDEXING_BLOCK_SIZE = 100;
+
+    /** @var Client */
+    private $client;
+
+    /** @var LoggerInterface */
+    private $logger;
+
+    public function __construct(Client $client, LoggerInterface $logger)
+    {
+        $this->client = $client;
+        $this->logger = $logger;
+    }
+
+    public function getIndexNameByDocument(IndexDocument $document): string
+    {
+        $documentBody = $document->getBody();
+
+        if (!isset($documentBody['type'])) {
+            throw new RuntimeException('type property is undefined on the document');
+        }
+
+        $documentType = is_string($documentBody['type']) ? [$documentBody['type']] : $documentBody['type'];
+
+        foreach (self::AVAILABLE_INDEXES as $ontology => $indexName) {
+            if (in_array($ontology, $documentType)) {
+                return $indexName;
+            }
+        }
+
+        return self::UNCLASSIFIEDS_DOCUMENTS_INDEX;
+    }
+
+    /**
+     * @param Iterator $documents
+     * @return int The number of indexed documents
+     */
+    public function buildIndex(Iterator $documents): int
+    {
+        $count = 0;
+        $blockSize = 0;
+        $params = [];
+
+        while ($documents->valid()) {
+            /** @var IndexDocument $document */
+            $document = $documents->current();
+
+            // First step we trying to create document. If is exist, then skip this step
+            $indexName = $this->getIndexNameByDocument($document);
+
+            $this->logger->info('indexname:' . $indexName);
+
+            if ($indexName === self::UNCLASSIFIEDS_DOCUMENTS_INDEX) {
+                $this->logger->warning(
+                    sprintf(
+                        'There is no proper index for the document "%s", type:"%s"',
+                        $document->getId(),
+                        var_export($document->getBody()['type'] ?? null, true)
+                    )
+                );
+                $documents->next();
+                continue;
+            }
+
+            $this->logger->info(sprintf('adding document "%s" to be indexed', $document->getId()));
+
+            $params = $this->extendBatch('index', $indexName, $document, $params);
+
+            $documents->next();
+
+            $blockSize++;
+
+            if ($blockSize === self::INDEXING_BLOCK_SIZE) {
+                $clientResponse = $this->client->bulk($params);
+
+                $this->logger->debug('client response: '. json_encode($clientResponse));
+
+                $count += $blockSize;
+                $blockSize = 0;
+                $params = [];
+            }
+        }
```

Comment:
Suggestion: Using `foreach` here might make it a bit easier to follow. All branches inside this loop end up calling `$documents->next();` so I think it would be safe to use `foreach` and remove the `next()` calls.

```suggestion
        foreach ($documents as $document) {
            /** @var IndexDocument $document */

            // First step we trying to create document. If is exist, then skip this step
            $indexName = $this->getIndexNameByDocument($document);

            $this->logger->info('indexname:' . $indexName);

            if ($indexName === self::UNCLASSIFIEDS_DOCUMENTS_INDEX) {
                $this->logger->warning(
                    sprintf(
                        'There is no proper index for the document "%s", type:"%s"',
                        $document->getId(),
                        var_export($document->getBody()['type'] ?? null, true)
                    )
                );

                continue;
            }

            $this->logger->info(sprintf('adding document "%s" to be indexed', $document->getId()));

            $params = $this->extendBatch('index', $indexName, $document, $params);

            $blockSize++;

            if ($blockSize === self::INDEXING_BLOCK_SIZE) {
                $clientResponse = $this->client->bulk($params);

                $this->logger->debug('client response: '. json_encode($clientResponse));

                $count += $blockSize;
                $blockSize = 0;
                $params = [];
            }
        }
```